api/src/services/stream.service.ts

```diff
@@ -0,0 +1,174 @@
+import { context, propagation, SpanKind, SpanStatusCode } from '@opentelemetry/api';
+import { SemanticAttributes, MessagingOperationValues, MessagingDestinationKindValues } from '@opentelemetry/semantic-conventions';
+import { InstrumentedComponent } from '@pokemon/telemetry/instrumented.component';
+import { createSpanFromContext, runWithSpan } from '@pokemon/telemetry/tracing';
+import { Kafka, Consumer, KafkaMessage, IHeaders } from 'kafkajs'
+import { CustomTags } from '../constants/Tags';
+
+const { KAFKA_BROKER = '', KAFKA_TOPIC = '', KAFKA_CLIENT_ID = '' } = process.env;
+
+export interface StreamingService<T> {
+  subscribe(callback: Function): Promise<void>;
+}
+
+function createStreamingService<T>(): StreamingService<T> {
+  const stream = new KafkaStreamService(KAFKA_TOPIC);
+  return new InstrumentedKafkaStreamService(KAFKA_TOPIC, stream);
+}
+
+class InstrumentedKafkaStreamService<T> extends InstrumentedComponent implements StreamingService<T> {
+  private readonly topic: string;
+  private readonly streamService: StreamingService<T>;
+
+  public constructor(topic: string, streamService: StreamingService<T>) {
+    super();
+    this.topic = topic;
+    this.streamService = streamService;
+  }
+
+  getBaseAttributes() {
+    return {
+      [SemanticAttributes.MESSAGING_SYSTEM]: 'kafka',
+      [SemanticAttributes.MESSAGING_DESTINATION]: this.topic,
+      [SemanticAttributes.MESSAGING_DESTINATION_KIND]: MessagingDestinationKindValues.TOPIC,
+    };
+  }
+
+  private extractHeaders(message: KafkaMessage) : IHeaders {
+    const kafkaHeaders = message.headers ?? {};
+    const headers = {};
+
+    // as Kafka send headers as buffers, we need to convert them to
+    // string, otherwise OTel SDK will not capture the trace context
+    for (let key in kafkaHeaders) {
+      let value = kafkaHeaders[key] || "";
+      if (value.toString !== undefined) {
+        value = value.toString();
+      }
+      headers[key] = value;
+    }
+
+    return headers
+  }
```

Comment:
I was expecting your usual wasted hours counter here :laughing: 

> This block of code owns me X hours of my life
> \- Daniel Dias

Reply:
Indeed 😄 !

This block of code owns me 2 days of my life!